# Ethel Esenam Klu

Data Cleaning Project
The purpose of data cleaning: 
1. Is to ensure that each column has the correct format of data
2. To check for duplicates in the dataset
3. Check for missing values and deal with it appropriately

This is very important because uncleaned data can adversely affect your data analysis

The first step is to import your packages

In [1]:
#importing the necessary packages 
import pandas as pd
import numpy as np
import re
import matplotlib.pyplot as plt
import seaborn as sns

The data used is an airbnb data which has been saved in a csv format and would have to be loaded into pandas and given a variable

In [2]:
#reading the csv data
airbnb = pd.read_csv('airbnb_data.csv')

In [3]:
#calling the dataframe gives all the rows and columns of that dataset
airbnb

id date_account_created  timestamp_first_active  \
0       gxn3p5htnn           2010-06-28          20090319043255   
1       820tgsjxq7           2011-05-25          20090523174809   
2       4ft3gnwmtx           2010-09-28          20090609231247   
3       bjjt8pjhuk           2011-12-05          20091031060129   
4       87mebub9p4           2010-09-14          20091208061105   
...            ...                  ...                     ...   
213446  zxodksqpep           2014-06-30          20140630235636   
213447  mhewnxesx9           2014-06-30          20140630235719   
213448  6o3arsjbb4           2014-06-30          20140630235754   
213449  jh95kwisub           2014-06-30          20140630235822   
213450  nw9fwlyb5f           2014-06-30          20140630235824   

       date_first_booking     gender   age signup_method  signup_flow  \
0                     NaN  -unknown-   NaN      facebook            0   
1                     NaN       MALE  38.0      facebook            0   
2              2010-08-02     FEMALE  56.0         basic            3   
3              2012-09-08     FEMALE  42.0      facebook            0   
4              2010-02-18  -unknown-  41.0         basic            0   
...                   ...        ...   ...           ...          ...   
213446                NaN       MALE  32.0         basic            0   
213447                NaN  -unknown-   NaN         basic            0   
213448                NaN  -unknown-  32.0         basic            0   
213449                NaN  -unknown-   NaN         basic           25   
213450                NaN  -unknown-   NaN         basic           25   

       language affiliate_channel affiliate_provider first_affiliate_tracked  \
0            en            direct             direct               untracked   
1            en               seo             google               untracked   
2            en            direct             direct               untracked   
3            en            direct             direct               untracked   
4            en            direct             direct               untracked   
...         ...               ...                ...                     ...   
213446       en         sem-brand             google                     omg   
213447       en            direct             direct                  linked   
213448       en            direct             direct               untracked   
213449       en             other              other           tracked-other   
213450       en            direct             direct               untracked   

       signup_app first_device_type  first_browser country_destination  
0             Web       Mac Desktop         Chrome                 NDF  
1             Web       Mac Desktop         Chrome                 NDF  
2             Web   Windows Desktop             IE                  US  
3             Web       Mac Desktop        Firefox               other  
4             Web       Mac Desktop         Chrome                  US  
...           ...               ...            ...                 ...  
213446        Web       Mac Desktop         Safari                 NDF  
213447        Web   Windows Desktop         Chrome                 NDF  
213448        Web       Mac Desktop        Firefox                 NDF  
213449        iOS            iPhone  Mobile Safari                 NDF  
213450        iOS            iPhone      -unknown-                 NDF  

[213451 rows x 16 columns]

In [4]:
#describe function gives an overview of the quantitative variables
airbnb.describe()

timestamp_first_active            age    signup_flow
count            2.134510e+05  125461.000000  213451.000000
mean             2.013085e+13      49.668335       3.267387
std              9.253717e+09     155.666612       7.637707
min              2.009032e+13       1.000000       0.000000
25%              2.012123e+13      28.000000       0.000000
50%              2.013091e+13      34.000000       0.000000
75%              2.014031e+13      43.000000       0.000000
max              2.014063e+13    2014.000000      25.000000

In [5]:
#info function outputs information about the data
airbnb.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 213451 entries, 0 to 213450
Data columns (total 16 columns):
 #   Column                   Non-Null Count   Dtype  
---  ------                   --------------   -----  
 0   id                       213451 non-null  object 
 1   date_account_created     213451 non-null  object 
 2   timestamp_first_active   213451 non-null  int64  
 3   date_first_booking       88908 non-null   object 
 4   gender                   213451 non-null  object 
 5   age                      125461 non-null  float64
 6   signup_method            213451 non-null  object 
 7   signup_flow              213451 non-null  int64  
 8   language                 213451 non-null  object 
 9   affiliate_channel        213451 non-null  object 
 10  affiliate_provider       213451 non-null  object 
 11  first_affiliate_tracked  207386 non-null  object 
 12  signup_app               213451 non-null  object 
 13  first_device_type        213451 non-null  object 
 14  firs

In [6]:
#printing out the first five rows of the dataset
airbnb.head()

id date_account_created  timestamp_first_active date_first_booking  \
0  gxn3p5htnn           2010-06-28          20090319043255                NaN   
1  820tgsjxq7           2011-05-25          20090523174809                NaN   
2  4ft3gnwmtx           2010-09-28          20090609231247         2010-08-02   
3  bjjt8pjhuk           2011-12-05          20091031060129         2012-09-08   
4  87mebub9p4           2010-09-14          20091208061105         2010-02-18   

      gender   age signup_method  signup_flow language affiliate_channel  \
0  -unknown-   NaN      facebook            0       en            direct   
1       MALE  38.0      facebook            0       en               seo   
2     FEMALE  56.0         basic            3       en            direct   
3     FEMALE  42.0      facebook            0       en            direct   
4  -unknown-  41.0         basic            0       en            direct   

  affiliate_provider first_affiliate_tracked signup_app first_device_type  \
0             direct               untracked        Web       Mac Desktop   
1             google               untracked        Web       Mac Desktop   
2             direct               untracked        Web   Windows Desktop   
3             direct               untracked        Web       Mac Desktop   
4             direct               untracked        Web       Mac Desktop   

  first_browser country_destination  
0        Chrome                 NDF  
1        Chrome                 NDF  
2            IE                  US  
3       Firefox               other  
4        Chrome                  US

Checking for duplicates in the dataset can be depicted in the cell below

In [10]:
airbnb_duplicates = airbnb.duplicated(keep=False)
airbnb_duplicates

0         False
1         False
2         False
3         False
4         False
          ...  
213446    False
213447    False
213448    False
213449    False
213450    False
Length: 213451, dtype: bool

In [ ]:
#checking the datatype of each column
airbnb.dtypes

The data types of the date_account_created and time_first_stamp columns should be integers instead of objects since we are dealing with dates and numbers here
The age column is also in a float format which is not the case as ages are always integers.
There has been some upcasting done to these columns, hence the need to change the data types

In [ ]:
#changing the data type of date_account_created
airbnb['date_account_created'] = pd.to_datetime(airbnb['date_account_created'])

In [ ]:
#checking the datatype 
airbnb['date_account_created'].dtype

In [ ]:
airbnb.info()

In [ ]:
#changing the data type
airbnb["timestamp_first_active"] = pd.to_datetime(airbnb["timestamp_first_active"],format = "%Y%m%d%H%M%S")

The data given sometimes contains missing values, so it is prudent to look for the missing values and decide what to do with them

In [ ]:
#checking for null or missing values in the dataset
airbnb.isnull()

Notice that the output gives False for areas with no Null values and True for areas with null values
The sum of null values in each column can be found as depicted in the cell below

In [ ]:
airbnb.isnull().sum()

The first_affiliate_tracked, date_first_booking and age columns all have missing values

In [ ]:
#checking the number of rows and columns in the dataframe 
airbnb.shape

The total number of rows in the first_affiliate_tracked column is 213451 of which 6065 of them are null

Since the null values do not make up at least 50% of the data, the column cannot be dropped. To fill the null values, the forward fill method can be used. Each null value will be replaced by the value that comes before it

In [ ]:
airbnb['first_affiliate_tracked'].head(15)

In [ ]:
#filling the missing values
airbnb['first_affiliate_tracked'].ffill(axis=None,inplace=True)

In [ ]:
airbnb['first_affiliate_tracked'].head(50)

In [ ]:
airbnb['first_affiliate_tracked'].isnull().sum()

The next column to be worked on is the date_first_booking column. It has 124543 out of 213451 being missing which is more than 50% of the data, consequently, it has to be dropped

In [ ]:
#dropping the column
airbnb.drop('date_first_booking', axis = 1, inplace = True)

In [ ]:
airbnb.head(0)

As seen in the above output, the column has been safely dropped

Age column is the next focus. The fillna method will be used here with the median of the values present. The column will be coverted to integer as well since we do not have floats for ages

In [ ]:
#filling missing values in the age column with the median of the values present
airbnb['age'] = airbnb['age'].fillna(airbnb['age'].median())

In [ ]:
#changing the datatype of the age column from a float to an integer
airbnb['age'] = airbnb['age'].astype('int')

In [ ]:
airbnb.head()

In [ ]:
airbnb['age'].isnull().sum()

In [ ]:
#stripping '-' under unknown in the gender column and making the column lowercase
airbnb['gender'] = airbnb['gender'].str.strip('-').str.lower()

In [ ]:
airbnb.head()

In [ ]:
airbnb['first_device_type'].unique()

In [ ]:
#stripping other from the other/unknown in the first_device_type column
airbnb['first_device_type'] = airbnb['first_device_type'].str.replace('Other/', ' ')

In [ ]:
airbnb['first_device_type'].unique()

In [ ]:
#stripping '-'
airbnb['first_browser'] = airbnb['first_browser'].str.strip('-')

In [ ]:
airbnb['first_browser'].unique()

In [ ]:
airbnb

The data is now clean and ready for analysis